ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitterSpec.scala

```diff
@@ -595,58 +595,134 @@ class TransactionCommitterSpec extends AnyWordSpec with Matchers with MockitoSug
+            Seq(false, true).map(inRollback =>
+              (name, newLookupByKeySubmittedTransaction(found, inRollback), contractIdAtCommitter)
             )
-        result shouldBe a[StepStop]
-        val rejectionReason =
-          getTransactionRejectionReason(result).getInconsistent.getDetails
-        rejectionReason should startWith("InconsistentKeys")
+          }
+
+      val casesTable = Table(
+        ("name", "transaction", "contractIdAtCommitter"),
+        cases: _*
+      )
+
+      forAll(casesTable) {
+        (_, transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
+          val context = commitContextWithContractStateKeys(
+            conflictingKey -> contractIdAtCommitter
+          )
+          val result = validate(context, transaction)
+          result shouldBe a[StepStop]
+          val rejectionReason =
+            getTransactionRejectionReason(result).getInconsistent.getDetails
+          rejectionReason should startWith("InconsistentKeys")
       }
     }
 
+    "return DuplicateKeys when two local contracts conflict" in {
+      val builder = TransactionBuilder()
+      builder.add(newCreateNodeWithFixedKey(s"#$freshContractId"))
+      builder.add(newCreateNodeWithFixedKey(s"#$freshContractId"))
+      val transaction = builder.buildSubmitted()
+      val context = commitContextWithContractStateKeys(conflictingKey -> None)
+      val result = validate(context, transaction)
+      result shouldBe a[StepStop]
+      val rejectionReason =
+        getTransactionRejectionReason(result).getInconsistent.getDetails
+      rejectionReason should startWith("DuplicateKeys")
+    }
+
+    "return DuplicateKeys when a create in a rollback conflicts with a global key" in {
+      val builder = TransactionBuilder()
+      val rollback = builder.add(builder.rollback())
+      builder.add(newCreateNodeWithFixedKey(s"#$freshContractId"), rollback)
+      val transaction = builder.buildSubmitted()
+      val context = commitContextWithContractStateKeys(conflictingKey -> Some(s"#$freshContractId"))
+      val result = validate(context, transaction)
+      result shouldBe a[StepStop]
+      val rejectionReason =
+        getTransactionRejectionReason(result).getInconsistent.getDetails
+      rejectionReason should startWith("DuplicateKeys")
+    }
+
+    "not return DuplicateKeys between local contracts if first create is rolled back" in {
+      val builder = TransactionBuilder()
+      val rollback = builder.add(builder.rollback())
+      builder.add(newCreateNodeWithFixedKey(s"#$freshContractId"), rollback)
+      builder.add(newCreateNodeWithFixedKey(s"#freshContractId"))
+      val transaction = builder.buildSubmitted()
+      val context = commitContextWithContractStateKeys(conflictingKey -> None)
+      val result = validate(context, transaction)
+      result shouldBe a[StepContinue[_]]
+    }
+
+    "return DuplicateKeys between local contracts even if second create is rolled back" in {
```

Comment:
Why this asymmetry though? What's the general principle for rollback validation w.r.t. keys here?

Reply:
Let’s say we only have creates in a rollback. In that case, you can imagine that instead of having a rollback we archive everything at the end of the rollback.

That means we have the two situations:
1. Rollback create first, then another create
This becomes
cid <- create
archive cid
create
The contract has been archived so no duplicate key
2. Create first, then rollback create
cid <- create
cid2 <- create
archive cid2
We archive cid2 here but we already fail on the second create with a duplicate key error.

Reply:
Thanks for the explanation. I can think of other test cases though, e.g.:

1. cid <- create with key k
2. consume cid
3. rollback 2
4. cid2 <- create with key k - SHOULD FAIL

Aren't they worth testing as well?

Reply:
Added the testcase in dd56491de514e356cdd1285b8cde46f617f27f82 I think there is a general question of how thorough you want to be wtr to testing here given that there were literally zero checks for duplicate keys before this PR but I like leaving things in a better state than they were before :slightly_smiling_face:

Reply:
I have only a partial view but I think that the lack of unit tests stems at least partially from a misunderstanding (or lack of understanding) about which component checks what.
I agree that it is beyond the scope of this PR to test the interaction with keys of constructs that predate rollback, but I think that tests about interactions of keys with rollback should instead be covered here as exhaustively as it makes sense (i.e., if something can possibly break individually, then it should have a dedicated unit test).
That is, unless we are going to very soon remove duplicated check logic in some way (e.g. b/c the engine is going to do it, or a shared lib for all to use will consolidate the logic).

Reply:
I do hope to do the engine-level checks very soon but regardless, I think the current state covers the cases I can think off (rolled-back create, rolled-back consuming choice, different orders). Do you have something else that you think is missing?

Reply:
If I understand correctly, only non-key-preserving actions need to be tested with rollback; consuming exercises are key-preserving, aren't they? If so, then testing archival should suffice. Another dimension is indeed global Vs. local; I am not sure if this can break individually but you've already tested it in other cases: how about also testing rolling back an archival of a conflicting global key?

Reply:
> consuming exercises are key-preserving, aren't they

I assume you mean non-consuming here? The archive is a consuming choice.

I added another test for a rolled-back archive of a global cid in 9db05395c906d41a50a8659a20e4fc5f99e46b8d